packages/utils/src/internals/recoverable-state.ts

```diff
@@ -0,0 +1,217 @@
+import { Configuration, EventType, KeyValueStore } from '@crawlee/core';
+
+import type { Log } from '@apify/log';
+import log from '@apify/log';
+
+export interface PersistenceOptions {
+    /**
+     * The key under which the state is stored in the KeyValueStore
+     */
+    persistStateKey: string;
+
+    /**
+     * Flag to enable or disable state persistence
+     */
+    persistenceEnabled?: boolean;
+
+    /**
+     * The name of the KeyValueStore to use for persistence.
+     * If neither a name nor an id are supplied, the default store will be used.
+     */
+    persistStateKvsName?: string;
+
+    /**
+     * The identifier of the KeyValueStore to use for persistence.
+     * If neither a name nor an id are supplied, the default store will be used.
+     */
+    persistStateKvsId?: string;
+}
+
+/**
+ * Options for configuring the RecoverableState
+ */
+export interface RecoverableStateOptions<TStateModel = Record<string, unknown>> extends PersistenceOptions {
+    /**
+     * The default state used if no persisted state is found.
+     * A deep copy is made each time the state is used.
+     */
+    defaultState: TStateModel;
+
+    /**
+     * A logger instance for logging operations related to state persistence
+     */
+    logger?: Log;
+
+    /**
+     * Configuration instance to use
+     */
+    config?: Configuration;
+
+    /**
+     * Optional function to transform the state to a JSON-serializable object before persistence.
+     * If not provided, the state is passed to JSON.stringify as-is.
+     */
+    serialize?: (state: TStateModel) => Record<string, unknown>;
```

Comment:
Shouldn't this do the JSON serialization as well? Or maybe it could even output into a stream so that we can update the value in KVS more efficiently.